contract/src/main/kotlin/io/provenance/scope/loan/utility/FundingDataValidation.kt

```diff
@@ -0,0 +1,106 @@
+package io.provenance.scope.loan.utility
+
+import tech.figure.loan.v1beta1.Funding
+import tech.figure.util.v1beta1.ACH
+
+internal fun Funding?.isSet() = this !== null && (status.isSet() || started.isSet() || completed.isSet() || disbursementsCount > 0)
+
+internal val fundingValidation: ContractEnforcementContext.(Funding) -> Unit = { funding ->
+    if (funding.isSet()) {
+        funding.status.takeIf { it.isSet() }?.also { fundingStatus ->
+            requireThat(
+                fundingStatus.status.isNotBlank()                   orError "Funding status must not be blank",
+                fundingStatus.effectiveTime.isValidForFundingTime() orError "Funding status must have valid effective time",
+            )
+        } ?: raiseError("Funding status must be set")
+        requireThat(
+            funding.started.isValidForFundingTime()   orError "Funding start time must be valid",
+            funding.completed.isValidForFundingTime() orError "Funding end time must be valid",
+        )
+        if (funding.disbursementsCount > 0) {
+            val incomingDisbursementIds = mutableMapOf<String, UInt>()
+            funding.disbursementsList.requireThatEach { disbursement ->
+                disbursement.takeIf { it.isSet() }?.also { setDisbursement ->
+                    setDisbursement.id.takeIf { it.isSet() }?.value?.also { disbursementId ->
+                        incomingDisbursementIds[disbursementId] = incomingDisbursementIds.getOrDefault(disbursementId, 0U) + 1U
+                    }
+                    requireThat(
+                        setDisbursement.id.isValid()                      orError "Disbursement must have valid ID",
+                        setDisbursement.started.isValidForFundingTime()   orError "Disbursement start time must be valid",
+                        setDisbursement.completed.isValidForFundingTime() orError "Disbursement end time must be valid",
+                        (setDisbursement.amount.value.toDoubleOrNull()?.let { it >= 0 } ?: true)
+                            orError "Disbursement amount must not be negative",
+                    )
+                    setDisbursement.status.takeIf { it.isSet() }?.also { disbursementStatus ->
+                        requireThat(
+                            (disbursementStatus.status in listOf("UNFUNDED", "INITIATED", "COMPLETED", "CANCELLED")) // TODO: Confirm if desired
```

Comment:
That is correct. It can be only be these 4 statuses in MAM. The P8e will still be the model in the domain layer for service-funding and confirmed the mapper mapped all the scenarios/states it can be in and in the correct order

[p8e-to-mam](https://cs.github.com/FigureTechnologies/provenance-corelib/blob/cd0ab7691864148d69cd27699b21d6ad11cb0af0/core-p8e-domain/src/main/kotlin/io/p8e/loan/P8eToMetadataAssetModelConverter.kt?q=UNFUNDED+repo%3AFigureTechnologies%2Fprovenance-corelib#L690)

Reply:
Sorry double checking this. I think these are funding status states. Let me see which status disbursement can be

Reply:
Based on the mapper:
- If the Funding Record is present it will use the Funding Status for the disbursement records
- If the Funding Record is not present it will use the fiatTransferDetail which shouldn't be populated since it is a deprecated field (processor-dorf doesn't currently populate this).

Since the funding record will always be present for this contract function, then I think matching the statuses with funding status is correct

Reply:
So just to be clear, since you say "it will use the Funding Status for the disbursement record statuses", is this (line 36), which checks each disbursement's status, _and_ [line 12](https://github.com/provenance-io/loan-package-contracts/pull/67/files#diff-7a4c19cc37adbe306cfb07d8b086ea2f8a4659af65a39e1cad238d6adf10e715R12), which checks the funding status, in accordance with what you expect? Because the code you linked above seems to be modifying the field that line 12 checks, not line 36.

Reply:
@rgipson-figure just tagging for visibility ^^

Reply:
Per discussion, moved this check to the top-level status field.